### PR TITLE
Allow calico/node to not override node resource parms if AS, IP or IP6 options are not specified

### LINKED ISSUE #1350: Update calicoctl node run to optionally overwrite the node resource parameters
Currently the node run command takes --as, --ip and --ip6 parameters to allow you to configure the AS number and IP addresses on the node.

However, these are also configurable through the node resource.

We should make it so that if the parameters are omitted then we will use the currently configured values.  If a node resource has not been explicitly created and set these values then we will:
-  Auto detect an IPv4 address
-  Use the global default AS number

### PR DESCRIPTION
**Pull request by robbrockbank**
Fixes #1350 

Requires #1319

### CODE REVIEW

File: calico_node/startup.py [hunk 1]
```diff
@@ -252,15 +252,19 @@ def main():
             print "Connected to etcd"
             break
 
+    # Query the currently configured host IPs
+    try:
+        ip, ip6 = client.get_host_bgp_ips(hostname)
```
Comment (caseydavenport) [inline calico_node/startup.py]: We should log out that we found an IP address in the datastore.

Comment (caseydavenport) [inline calico_node/startup.py]: I guess we've only got "print" here, so maybe that's too verbose...

File: calicoctl/commands/node/run.go [hunk 2]
```diff
@@ -181,15 +194,23 @@ Description:
 
 	// Create a mapping of environment variables to values.
 	envs := map[string]string{
-		"HOSTNAME": name,
```
Comment (caseydavenport) [inline calicoctl/commands/node/run.go]: The Felix runit job expects HOSTNAME to be set I think.  It sets the FELIX_HOSTNAME value based on that.

Comment (caseydavenport) [inline calicoctl/commands/node/run.go]: We might want to keep both around for a little while, so that users can use the new calicoctl to launch old calico/node images?

Comment (robbrockbank) [inline calicoctl/commands/node/run.go]: I don't believe we need to pass in HOSTNAME. 

The startup script honors both HOSTNAME and NODENAME and writes the name into a startup environments file that is used by rc.local  (and therefore by Felix and confd)

File: calico_node/startup.py [hunk 3]
```diff
@@ -252,39 +252,58 @@ def main():
...
-    if ip6 and not netaddr.valid_ipv6(ip6):
-        print "IP6 environment (%s) is not a valid IPv6 address." % ip6
-        sys.exit(1)
+    ip6 = os.getenv("IP6") or None
+    if ip6:
+        if not netaddr.valid_ipv6(ip6):
+            print "IP6 environment (%s) is not a valid IPv6 address." % ip6
+            sys.exit(1)
+        print "Using IPv6 address from IP6 environment: %s" % ip
+    elif current_ip:
```
Comment (caseydavenport) [inline calico_node/startup.py]: should be current_ip6

File: calico_node/startup.py [hunk 4]
```diff
@@ -280,11 +284,12 @@ def main():
...
             print "Couldn't autodetect a management IP address. Please " \
-                  "provide an IP address by rerunning the container with the" \
-                  " IP environment variable set."
+                  "provide an IP address either by configuring one in the " \
+                  "node resource, or by rerunning the container with the " \
+                  "IP environment variable set."
             sys.exit(1)
         else:
-            print "No IP provided. Using detected IP: %s" % ip
+            print "No IP configured or provided. Using detected IP: %s" % ip
```
Comment (robbrockbank) [inline calico_node/startup.py]: That is a valid fear - so I've added an "autodetect" option to the node run and to the container.

File: calico_node/startup.py (hunk 3 above)
Comment (robbrockbank) [inline calico_node/startup.py]: ha!  already fixed :-)

**Comment by caseydavenport** [comment]
Looking good to me.

Something about the auto-detect stuff makes me a little uneasy - it might be related to #873 though.

**Comment by robbrockbank** [comment]
Yes the auto-detect stuff makes me nervous too.

We could add a auto-discovery method option that indicates which auto-discovery method to use, but that does sound like a separate PR to me.

**Comment by caseydavenport** [comment]
yeah, definitely separate PR.  I just want to convince myself that an IP=auto option doesn't mean we're shooting ourselves in the foot somehow.

File: calicoctl/commands/node/run.go [hunk 5]
```diff
@@ -63,17 +63,29 @@ func Run(args []string) {
...
+     --as=<AS_NUM>         Set the AS number for this node.  If omitted, the
+                           node will use the currently configured value and
+                           you can use the node resource to modify the AS
+                           number.  If there is no configured value and the
+                           the --as option is omitted, the node will inherit
+                           the global AS number (see 'calicoctl config' for
+                           details).
+     --ip=<IP>             Set the local IPv4 routing address for this node.
+                           If omitted, the node will use the currently
+                           configured value and you can use the node resource
```
Comment (caseydavenport) [inline calicoctl/commands/node/run.go]: I'd reword this to "If omitted, will use the value configured on the node object."

**Comment by caseydavenport** [comment]
LGTM

File: calicoctl/commands/node/run.go [hunk 6]
```diff
@@ -65,27 +65,23 @@ Options:
...
-                           configured value and you can use the node resource
-                           to modify the address.  If there is no configured
-                           value and the --ip option is omitted, the node will
+                           If omitted, it will use the value configured on the
+                           node resource.  If there is no configured value
+                           and the --ip option is omitted, the node will
                            attempt to autodetect an IP address to use.  Use a
                            value of 'autodetect' to always force autodetection
-                           of the IP everytime the node starts.
+                           of the IP each the node starts.
```
Comment (gunjan5) [inline calicoctl/commands/node/run.go]: did you mean to write `each time` here?

File: calico_node/startup.py [hunk 7]
```diff
@@ -307,6 +289,31 @@ def main():
...
+    elif current_ip6:
+        print "Using configured IPv6 address: %s" % current_ip6
+        ip6 = current_ip6
+    else:
+        print "No IPv6 address configured"
+
+    as_num = os.getenv("AS")
+    if as_num:
+        if not validate_asn(as_num):
+            print "AS environment (%s) is not an AS number." % as_num
```
Comment (gunjan5) [inline calico_node/startup.py]: is not *a valid* AS number
